docs(sdk): document the compatibility layer for nextjs (#51173)
metabase/metabase: docs/embedding/sdk/next-js.md

Some notes on using the Embedded analytics SDK with [Next.js](https://nextjs.org/). The SDK is tested to work with Next.js 14, although it may work with other versions.

## SDK components with Server Side Rendering (SSR) or React Server Components

For now, the SDK components are only supported for client-side rendering. To use the SDK components with server-side rendering, or with React Server components, you can either use a compatibility layer or manually wrap the components.

### Compatibility layer for server-side rendering (SSR) (EXPERIMENTAL)

To use SDK components with Next.js, the SDK provides an experimental compatibility layer that [wraps all the components with dynamic imports and disables SSR](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-no-ssr). To work with the app router, this compatibility layer uses `use client`.

To use the compatibility layer, change your imports from `@metabase/embedding-sdk-react` to `@metabase/embedding-sdk-react/nextjs`.

> See a  [sample Next.js app that uses this compatibility layer](https://github.com/metabase/metabase-nextjs-sdk-embedding-sample).

Reviewer: ```suggestion
See a [sample Next.js app that uses this compatibility layer](https://github.com/metabase/metabase-nextjs-sdk-embedding-sample).
```

Some notes on using the Embedded analytics SDK with [Next.js](https://nextjs.org/). The SDK is tested to work with Next.js 14, although it may work with other versions.

> ## Making the SDK components work on Server Side Rendering (SSR) or React Server Components

Reviewer 2: ```suggestion
## SDK components with Server Side Rendering (SSR) or React Server Components
```

> Currently, the SDK components only work on the browser, this means that using them directly from Server Side Rendering (SSR) or Server Components is not supported.
> To make them work, you can either use the compatibility layer or manually wrap the components.

Reviewer 2: ```suggestion
For now, the SDK components are only supported for client-side rendering. To use the SDK components with server-side rendering, or with React Server components, you can either use a compatibility layer or manually wrap the components.
```

> ### Compatibility layer

Reviewer 2: ```suggestion
### Compatibility layer for server-side rendering (SSR) (EXPERIMENTAL)
```

> On the latest releases of the sdk do provide an experimental compatibility layer that wraps all the components with [dynamic imports and disable SSR](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-no-ssr) for them. The compatibility layer also uses `use client` so that it will work with the app router.
> 
> To use the compatibility layer you can change your imports from `@metabase/embedding-sdk-react` to `@metabase/embedding-sdk-react/nextjs`.
> 
> This is the easiest way to get up and running with the SDK in Next.js, you can find [sample apps using this approach on github](https://github.com/metabase/metabase-nextjs-sdk-embedding-sample).

Reviewer 2: ```suggestion
To use SDK components with Next.js, the SDK provides an experimental compatibility layer that [wraps all the components with dynamic imports and disables SSR](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-no-ssr). To work with the app router, this compatibility layer uses `use client`.

To use the compatibility layer, change your imports from `@metabase/embedding-sdk-react` to `@metabase/embedding-sdk-react/nextjs`.

See a  [sample Next.js app that uses this compatibility layer](https://github.com/metabase/metabase-nextjs-sdk-embedding-sample).
```

## Manual wrapping of the components

> If the compatibility layer doesn't work for your use case, for example because you want to customize the loading of the components, you can create your own wrapper.
> Here's a suggested way to do it.
> 
> First, create a `metabase` folder, and inside it create a file created `EmbeddingSdkProvider.tsx`.
> This file will contain the provider with the appropriate configuration.

Reviewer 2: ```suggestion
If you want to customize the loading of the components, you can create your own wrapper.

In your app, create a `metabase` directory, and add a `EmbeddingSdkProvider.tsx` file to that directory. This file will contain the provider with the appropriate configuration.
```

```tsx
"use client";

import {
  defineMetabaseAuthConfig,
  MetabaseProvider,
} from "@metabase/embedding-sdk-react";

const authConfig = defineMetabaseAuthConfig({
  metabaseInstanceUrl: process.env.NEXT_PUBLIC_METABASE_INSTANCE_URL,
  authProviderUri: process.env.NEXT_PUBLIC_METABASE_AUTH_PROVIDER_URI,
});

export const EmbeddingSdkProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  return (
    <MetabaseProvider authConfig={authConfig}>{children}</MetabaseProvider>
  );
};
```

> Then, we need to create a `index.tsx` file that will export a lazy loaded version, with SSR disabled, and with the `use client` directive, of the `EmbeddingSdkProvider` component and all the other components you need.

Reviewer 2: ```suggestion
Next, add an `index.tsx` file to that `metabase` directory. This file will include the `use client` directive, and it'll export a lazy-loaded version of the `EmbeddingSdkProvider` with  SSR disabled.
```

```tsx
"use client";

import dynamic from "next/dynamic";

import React from "react";

// Lazy load the EmbeddingSdkProvider so and let it render children while it's being loaded
export const EmbeddingSdkProviderLazy = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const EmbeddingSdkProvider = dynamic(
    () =>
      import("./EmbeddingSdkProvider").then(m => {
        return { default: m.EmbeddingSdkProvider };
      }),
    {
      ssr: false,
      loading: () => {
        // render children while loading
        return <div>{children}</div>;
      },
    },
  );

  return <EmbeddingSdkProvider>{children}</EmbeddingSdkProvider>;
};

// Wrap all components that you need like this:

export const StaticQuestion = dynamic(
  () => import("@metabase/embedding-sdk-react").then(m => m.StaticQuestion),
  {
    ssr: false,
    loading: () => {
      return <div>Loading...</div>;
    },
  },
);

export const StaticDashboard = dynamic(
  () => import("@metabase/embedding-sdk-react").then(m => m.StaticDashboard),
  {
    ssr: false,
    loading: () => {
      return <div>Loading...</div>;
    },
  },
);
```

> You can now use the components in your pages or components like this:

Reviewer 2: ```suggestion
You can now import components like so:
```
